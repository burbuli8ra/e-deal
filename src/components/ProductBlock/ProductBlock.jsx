import React, { useState } from 'react';
import Styled from './ProductBlock.style';

const ProductBlock = ({
  id,
  imageUrl,
  price,
  title
}) => {
  const [isImageLoading, setIsImageLoading] = useState(true);
  const [hasVerticalImage, setHasVerticalImage] = useState(false);

  const priceInEuros = (price/100).toFixed(2);

  const handleImageLoad = ({ currentTarget }) => {
    const { height, width } = currentTarget.getBoundingClientRect();
    if (height > width) {
      setHasVerticalImage(true);
    }
    setIsImageLoading(false);
  };

  return (
    <>
      {isImageLoading ? (
        <Styled.GhostImage
          alt={title}
          data-testid="productItem-ghost-image"
          onLoad={handleImageLoad}
          src={imageUrl}
        />
      ) : (
        <Styled.ProductBlock data-testid="productItem-block" to={`/products/${id}`}>
          <Styled.ImageWrap>
            <Styled.Image
              alt={title}
              isVertical={hasVerticalImage}
              src={imageUrl}
            />
          </Styled.ImageWrap>
          <Styled.Title>{title}</Styled.Title>
          <Styled.Price>{priceInEuros}â‚¬</Styled.Price>
        </Styled.ProductBlock>
      )}
    </>
  );
};

export default ProductBlock;