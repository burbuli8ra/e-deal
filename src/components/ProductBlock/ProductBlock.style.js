import styled from '@emotion/styled';
import { Link } from 'react-router-dom';
import { colors } from 'theme';

const ProductBlock = styled(Link)`
  align-self: flex-start;
  background-color: ${colors.white};
  border-radius: 4px;
  box-shadow: 0 0 9px 0 rgba(140, 140, 158, 0.25);
  color: ${colors.pastelBlue};
  display: inline-flex;
  flex: 1 0 auto;
  flex-direction: column;
  margin: 10px;
  max-width: 240px;
  overflow: hidden;
  padding: 20px;
  text-decoration: none;
`;
ProductBlock.displayName = 'ProductBlock';

const GhostImage = styled.img`
  overflow: hidden;
  opacity: 0;
  visibility: hidden;
`;
GhostImage.displayName = 'GhostImage';

const Image = styled.img`
  margin-bottom: 10px;
  
  ${({ isVertical }) =>
    isVertical ? `
      height: 180px;
      width: auto;
    ` : `
      height: auto;
      width: 180px;
    `}
`;
Image.displayName = 'Image';

const ImageWrap = styled.div`
  align-items: center;
  display: flex;
  height: 180px;
  width: 100%;
  justify-content: center;
`;
ImageWrap.displayName = 'ImageWrap';

const Price = styled.div`
  font-size: 18px;
  font-weight: 700;
`;
Price.displayName = 'Price';

const Title = styled.h3`
  font-size: 16px;
  font-weight: 400;
  height: 38px;
  margin-bottom: 20px;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
`;
Title.displayName = 'Title';

export default {
  ProductBlock,
  GhostImage,
  Image,
  ImageWrap,
  Price,
  Title
};