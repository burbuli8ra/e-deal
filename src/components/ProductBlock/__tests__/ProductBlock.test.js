import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import Mocks from '__mocks__';
import ProductBlock from '../ProductBlock';

const setup = props => (
  <Router>
    <ProductBlock {...props} />
  </Router>
);

describe('ProductBlock component', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = Mocks.ProductItem;
    wrapper = render(setup(props));
  });

  test('should match product ghost image snapshot', () => {
    const { asFragment } = wrapper;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should match product block snapshot', () => {
    const { asFragment, getByTestId } = wrapper;
    fireEvent.load(getByTestId('productItem-ghost-image'));
    expect(asFragment()).toMatchSnapshot();
  });
});