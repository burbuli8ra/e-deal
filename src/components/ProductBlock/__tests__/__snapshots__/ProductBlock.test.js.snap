// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBlock component should match product block snapshot 1`] = `
<DocumentFragment>
  <a
    class="css-1sj3ugd"
    data-testid="productItem-block"
    href="/products/2155431109"
  >
    <div
      class="css-1yv013g"
    >
      <img
        alt="Hobby HSM-1000 White"
        class="css-y6bwx0"
        src="https://abpcdn.pstatic.gr/P/bpimg34/2vgIi/1JiA8P_SX910Y902/0/2155431109.webp"
      />
    </div>
    <h3
      class="css-15l8jez"
    >
      Hobby HSM-1000 White
    </h3>
    <div
      class="css-g0fl43"
    >
      61.90â‚¬
    </div>
  </a>
</DocumentFragment>
`;

exports[`ProductBlock component should match product ghost image snapshot 1`] = `
<DocumentFragment>
  <img
    alt="Hobby HSM-1000 White"
    class="css-vh3imo"
    data-testid="productItem-ghost-image"
    src="https://abpcdn.pstatic.gr/P/bpimg34/2vgIi/1JiA8P_SX910Y902/0/2155431109.webp"
  />
</DocumentFragment>
`;
